### ISSUE TITLE
Parsing with incomplete grammar

### ISSUE THREAD

**Issue by chrullrich** [lark-parser/lark#214]
Hello,

I want to parse a voluminous line-based input, but I am only interested in a fairly small subset of the lines appearing in the file. Everything else should be ignored as silently as possible.

As an example, take this input:

    alpha
    beta
    gamma
    delta
    epsilon

Of these five lines, I am only interested in, let's say, the first ~two~ three. From the documentation and existing issues on the subject, it appears that rule priorities and a `/.*/` catch-all rule are the best approach. It appears, however, that this does not always work correctly. 

My best effort to date is:

    start: (line | _NL)+

    line: alpha | beta | other

    alpha.1: "alpha"
    beta.2: "beta"
    other.100: /.+/

    %import common.NEWLINE -> _NL
    %import common.WS_INLINE
    %ignore WS_INLINE

The result of this grammar is dependent on PYTHONHASHSEED; testing with [1..100] the results for [1..9] are correct; all others only identify "beta".

After adding the "gamma" rule (`line: alpha | beta | gamma | other`, `gamma.3: "gamma"`), I get 13 correct results, 9 with only "beta", and 78 where _all_ lines are "other".

With the "dynamic_complete" lexer I found mentioned somewhere, the results are identical, and with `ambiguity="explicit"` added I get a ton of alternatives that apparently differ only in where the "epsilon" is cut in half ("epsilon", "e"/"psilon", etc.). The "beta" effect of PYTHONHASHSEED is still there, too.

Is there a reliable way to do this? I could prefilter the input, but by that point I'm not very far from rolling my own ad-hoc parser, and I'd rather avoid that.

**Comment by erezsh** [comment]
This sort of highly-ambiguous parsing stretches the existing implementation, which is admitted flawed.

Please give the `earley_sppf` branch a try (https://github.com/lark-parser/lark/tree/earley_sppf). I'd be very interested to hear if it improves the situation in any way.

**Comment by chrullrich** [comment]
`parser="earley", lexer="standard"` on that branch, right? (With the "dynamic" lexer I _always_ get "other".)

This gives me the correct result for the test case above, but as soon as I add more tokens to the rules, things go south. Parsing

    alpha one
    beta
    gamma
    delta
    epsilon

with

    start: (line | _NL)+

    line: alpha | beta | gamma | other

    alpha: "alpha" WORD
    beta: "beta" WORD
    gamma: "gamma" WORD
    other: /.+/

    %import common.WORD
    %import common.NEWLINE -> _NL
    %import common.WS_INLINE
    %ignore WS_INLINE

results in

```
Traceback (most recent call last):
  File "testlark.py", line 23, in <module>
    tree = parser.parse(open("test.txt", "rt").read())
  File "c:\daten\projekte\other\lark\lark\lark.py", line 222, in parse
    return self.parser.parse(text)
  File "c:\daten\projekte\other\lark\lark\parser_frontends.py", line 36, in parse
    return self.parser.parse(token_stream, *[sps] if sps is not NotImplemented else [])
  File "c:\daten\projekte\other\lark\lark\parsers\earley.py", line 185, in parse
    column, to_scan = scan(i, token, column, to_scan)
  File "c:\daten\projekte\other\lark\lark\parsers\earley.py", line 152, in scan
    raise UnexpectedToken(token, expect)
lark.exceptions.UnexpectedToken: Unexpected token Token(_NL, '\n') at line 2, column 5.
Expected: WORD
```

I expected the parser to fall back (there's probably a better word for that) to "other" when it cannot match the "beta" rule.

**Comment by erezsh** [comment]
> I expected the parser to fall back  to "other" 

Actually, that's exactly what the `standard` lexer doesn't do.

Your best bet is to get the dynamic lexer to work. Perhaps if you provide some minimal example, I can give it a try.

**Comment by erezsh** [comment]
Oh, btw

> With the "dynamic" lexer I always get "other"

That might be because you gave it a priority of 100, which (unless I'm confused) means the highest priority.

**Comment by chrullrich** [comment]
For the test case, that is. The time behavior of the parser is an absolute killer; run time is something like O(1.85^N) [sic!] in the number of lines, For my expected input that works out to quite a few times the age of the universe.

I suspect this is because there is too much lookahead, but I've been fiddling with _NL tokens here and there to give the parser a chance to detect end-of-line, with no success whatsoever.

[testlark.py.txt](https://github.com/lark-parser/lark/files/2314462/testlark.py.txt)

**Comment by erezsh** [comment]
If you know your input is always divided by newlines, why not split on newlines and input each line separately? Unless there is structure in the way the lines are organized?

If there is any way to limit the time `other` is matched (right now, it's matched for every line), that will also improve performance.

If none of these work (I don't mind working through a few more ideas), perhaps it's one of those rare situations in which PEGs would be more suitable.

**Comment by chrullrich** [comment]
Yes, the lines _are_ independent, and splitting by newlines works. Avoiding "other" matches is not that easy, because I keep getting nondeterministic (i.e. PYTHONHASHSEED-dependent) behavior, where the same line works sometimes, but not all the time:

    start: (instruction | other | _NL)+

    instruction: beta _NL

    beta: "beta" beta_option

    !beta_option: "add" _name
               | beta_other

    other: /(?!beta)[^\n]+/ _NL
    beta_other: /(?!add).+/

    _name: _QUOTE CNAME _QUOTE
         | CNAME

    _QUOTE: /\"/

    %import common.CNAME
    %import common.NEWLINE -> _NL
    %import common.WS_INLINE
    %ignore WS_INLINE

As far as I understand it, this should work without priorities because the "beta other-than-add" case is excluded by the negative lookahead assertion. However:

```
[input is "\n".join(["beta add something"] * 4)]

start
  instruction
    beta
      beta_option
        beta_other       add something
  instruction
    beta
      beta_option
        beta_other       add something
  instruction
    beta
      beta_option
        add
        something
  instruction
    beta
      beta_option
        add
        something
```

In the "beta_other" cases, there is a space before "add".

**Comment by whitten** [comment]
Are you choosing the lines to ignore based on their location in the file or based on some characteristic (text or otherwise) of the contents of the line?
If it is based solely on location, perhaps we should have the ability of a rule to match the line number in the file. perhaps some meta syntax that means "the current line number is blah".  This would be an extension to EBNF, but I can see it being very useful in certain parsing situations. I'm actually surprised that I haven't heard this suggested before, but it is very useful.

If the match is based on the contents of the line, you are going to have to at least look at the line enough to reject it.  Does Lark compute a FIRST set of all tokens that could appear at the front of each rule in the grammar ? Testing against that would be the fastest method, I think.

**Comment by chrullrich** [comment]
Based on each individual line's content. The actual input is a configuration file for a network switch (ExtremeXOS config is stateless, so no configuration modes like e.g. Cisco has; each line is one independent command). Most of the lines are some variation on "configure [this] [that way]", but I only want two or three [this]es.

A somewhat random sample of lines:

    create vlan "FACTORY"
    configure vlan FACTORY tag 1000
    disable flow-control rx-pause port 1
    configure vlan FACTORY add ports 1-15, 17, 32
    configure iproute add default 172.16.1.1
    disable port 16

Of these lines, I don't want the third and fifth, but I need the others. As I mentioned far above, I could pre-filter, i.e. feed only the lines I know I want to the parser, but once I have analyzed each line thorougly enough to *know* whether I want it, I have already extracted the necessary information as well. The other way around, excluding the lines I don't want, would be the same if done outside the parser, but in the parser I thought I could use negative lookahead.

As far as useful features for parsers go, I would *much* prefer having a clean way to say "when you reach a point where the input fits no known rule, don't panic, but skip ahead past the next newline and try again from there". I suppose this is much more difficult to do than it sounds, however.
